File: src/advent-of-code/2021/day-4/main.py
from pathlib import Path

INPUT_FILE_PATH = "./input.txt"
EXAMPLE_FILE_PATH = "./example.txt"

class Tile:
  def __init__(self, value: str):
    self.value = int(value)
    self.is_crossed = False

  def cross(self):
    self.is_crossed = True
  
  def handle_tile(self, row_index: int, tile_index: int, row_counters: list[int], column_counters: list[int]):
    self.cross()
    column_counters[tile_index] += 1
    row_counters[row_index] += 1
    if column_counters[tile_index] == 5 and row_counters[row_index] == 5:
      return True, "Cross!"
    if column_counters[tile_index] == 5:
      return True, "Column"
    if row_counters[row_index] == 5:
      return True, "Row"
    return False, "None"

def get_file_lines(file_path: str):
  file = Path(__file__).with_name(file_path)
  return file.read_text().splitlines()

def get_rounds_and_board(lines: list[str]):
  rounds = [int(char) for char in lines[0].split(",")]
  boards: list[list[list[Tile]]] = []
  is_collecting_board = False
  for index in range(1, len(lines)):
    line = lines[index]
    if is_collecting_board:
      board = boards[-1]
      row = [Tile(char) for char in line.split()]
      board.append(row)
      if len(board) == 5:
        is_collecting_board = False
    if line == "":
      is_collecting_board = True
      boards.append([])
  return rounds, boards

def create_board_counters(length: int):
  columns: dict[int, list[int]] = {}
  for index in range(0, length):
    columns[index] = [0 for i in range(0, 5)]
  return columns

def get_last_winner_board(file_path: str):
  lines = get_file_lines(file_path)
  rounds, boards = get_rounds_and_board(lines)
  columns = create_board_counters(len(boards))
  rows = create_board_counters(len(boards))
  completed_boards: dict[int, tuple[int, str]] = {}
  last_board_index = len(boards) - 1
  for number in rounds:
    for index, board in enumerate(boards):
      is_board_completed = index in completed_boards
      if is_board_completed: continue 
      column_counters = columns[index]
      row_counters = rows[index]
      for row_index, row in enumerate(board):
        if is_board_completed: break
        for tile_index, tile in enumerate(row): # column
          if tile.value == number:
            is_already_crossed = tile.is_crossed
            if is_already_crossed: continue
            has_won, w_line = tile.handle_tile(row_index, tile_index, row_counters, column_counters)
            if has_won:
              last_board_index = index
              is_board_completed = True 
              completed_boards[index] = tile.value, w_line
              break
  if last_board_index in completed_boards:
    return boards[last_board_index], *completed_boards[last_board_index]
  raise ValueError("Unable to Find Last Winner Board")

def get_first_winner_board(file_path: str):
  lines = get_file_lines(file_path)
  rounds, boards = get_rounds_and_board(lines)
  columns = create_board_counters(len(boards))
  rows = create_board_counters(len(boards))
  # bold of you to assume i would make this performant first try
  for number in rounds:
    for index, board in enumerate(boards):
      column_counters = columns[index]
      row_counters = rows[index]
      for row_index, row in enumerate(board):
        for tile_index, tile in enumerate(row): # column
          if tile.value == number:
            is_already_crossed = tile.is_crossed
            if is_already_crossed: continue
            has_won, w_line = tile.handle_tile(row_index, tile_index, column_counters, row_counters)
            if has_won:
              return board, tile.value, w_line
  raise ValueError("Unable to Find Winner Board")

def get_winner_board_score(file_path: str, fn = get_first_winner_board):
  w_board, w_number, w_line = fn(file_path)
  uncrossed_count = 0
  for row in w_board:
    for tile in row:
      if not tile.is_crossed:
        uncrossed_count += tile.value
  score = uncrossed_count * w_number
  return score, w_number, w_line

if __name__ == "__main__":
  score, w_number, w_line = get_winner_board_score(EXAMPLE_FILE_PATH)
  print("First Board Winner:")
  print(" -- Results from Example File:")
  print(f" ---- Winner Board Score: {score} | Winner Number: {w_number} | Won in: {w_line}")
  score, w_number, w_line = get_winner_board_score(INPUT_FILE_PATH)
  print(" -- Results from Input File:")
  print(f" ---- Winner Board Score: {score} | Winner Number: {w_number} | Won in: {w_line}")
  score, w_number, w_line = get_winner_board_score(EXAMPLE_FILE_PATH, get_last_winner_board)
  print("Last Board Winner:")
  print(" -- Results from Example File:")
  print(f" ---- Last Winner Board Score: {score} | Winner Number: {w_number} | Won in: {w_line}")
  score, w_number, w_line = get_winner_board_score(INPUT_FILE_PATH, get_last_winner_board)
  print(" -- Results from Input File:")
  print(f" ---- Last Winner Board Score: {score} | Winner Number: {w_number} | Won in: {w_line}")

File: src/advent-of-code/2021/day-4/main.rb
class Board
  attr_reader :rows

  def initialize 
    @rows = []
    @column_counters = Array.new(5, 0)
  end

  def get_board_score win_number
    uncrossed_count = @rows.reduce(0) do |count, row|
      count + row.tiles.reduce(0) do |count, tile|
        count + (tile.is_crossed ? 0 : tile.value)
      end
    end
    uncrossed_count * win_number
  end

  def cross row_index, tile_index
    has_won_with_row = @rows[row_index].cross(tile_index)
    @column_counters[tile_index] += 1
    column_count = @column_counters[tile_index]
    has_won_with_column = column_count == 5
    if has_won_with_row and has_won_with_column
      return true, "Cross!"
    end
    if has_won_with_row
      return has_won_with_row, "Row"
    end
    if has_won_with_column
      return true, "Column"
    end
    return false, "None"
  end

  def push row
    @rows.push(row)
  end

  def is_full
    @rows.length == 5
  end
end

class Row
  attr_reader :tiles

  def initialize line
    @count = 0
    @tiles = Row.parse_line(line)
    @is_completed = false
  end

  def cross tile_index
    tile = @tiles[tile_index]
    tile.cross
    @count += 1
    self.has_won
  end

  def has_won
    @count == 5
  end 

  def self.parse_line line
    line
      .gsub(/\s/, " ")
      .split(" ")
      .map { |char| Tile.new(char.to_i) }
  end
end

class Tile
  attr_reader :value
  attr_reader :is_crossed

  def initialize value
    @value = value
    @is_crossed = false
  end

  def cross
    @is_crossed = true
  end
end

def get_boards_and_rounds lines
  boards = []
  rounds = lines[0].split(",").map { |char| char.to_i }
  is_collecting_board = false
  for line in lines
    if is_collecting_board
      board = boards.last
      board.push(Row.new(line))
      if board.is_full
        is_collecting_board = false 
      end
    end
    if line.strip == ""
      is_collecting_board = true
      boards.push(Board.new)
    end
  end
  return boards, rounds
end

def get_first_board_winner lines
  boards, rounds = get_boards_and_rounds(lines)
  for round in rounds
    for board in boards
      board.rows.each.with_index do |row, row_index|
        row.tiles.each.with_index do |tile, tile_index|
          is_already_crossed = tile.is_crossed
          next if tile.value != round or is_already_crossed
          has_won, win_text = board.cross(row_index, tile_index)
          return board, tile.value, win_text if has_won 
        end
      end
    end
  end
end

def get_file_lines file_path
  IO.readlines(file_path)
end

def print_results file_name, score, win_number, win_line
  puts " -- Results from #{file_name}:"
  puts " ---- Winner Board Score #{score} | Winner Number #{win_number} | Won in: #{win_line}"
end

EXAMPLE_FILE_PATH ="./example.txt"
INPUT_FILE_PATH = "./input.txt"

lines = get_file_lines(EXAMPLE_FILE_PATH)
board, win_number, win_line = get_first_board_winner(lines)
score = board.get_board_score(win_number)
puts "First Board Winner"
print_results("Example File", score, win_number, win_line)
lines = get_file_lines(INPUT_FILE_PATH)
board, win_number, win_line = get_first_board_winner(lines)
score = board.get_board_score(win_number)
print_results("Input File", score, win_number, win_line)
